### ISSUE TITLE
Why serialize newtype struct as array?

### ISSUE THREAD

**Issue by SX91** [3Hren/msgpack-rust#129]
https://github.com/3Hren/msgpack-rust/blob/master/rmp-serde/src/encode.rs#L375

Newtype is simply a wrapper around inner data, so it should encode/decode as inner data. You shouldn't really treat it as one-field struct.

**Comment by 3Hren** [comment]
Since there is no standard description how these types should be serialized using various encoders each crate does it on its own.

The best thing I can offer is a some kind of configurable `Serializer`.

**Comment by SX91** [comment]
@3Hren newtype struct is a special (type-system only) case, almost like in Haskell with it's `newtype` keyword. It's a type-system level wrapper around some other type. So it should be treated as a special case. `serialize_newtype_struct` has been added to Serde exactly for this purpose.

Consider some example:
```rust
struct MyVecWrapper(Vec<u32>);
enum MyEnum {
    Variant1(MyVecWrapper),
    Variant2(MyVecWrapper, MyVecWrapper),
}
```

`MyEnum::Variant1(MyVecWrapper(...))``` would be serialized as `[0, [[[...]]]]`, wouldn't it? That's quite a lot of overhead, while it could be simple `[0, [...]]` like with Haskell's `data-msgpack` package.

**Comment by 3Hren** [comment]
I got the point, thanks. Still, I need to check whether some of my crates has already tied with pack-as-array behaviour.

**Comment by SX91** [comment]
@3Hren could you please explain https://github.com/3Hren/msgpack-rust/blob/master/rmpv/src/ext/de.rs#L717 ?
newtype_variant is `enum A {A(i32)}`, right? So there's no way there could be `T...`, isn't it?

**Comment by 3Hren** [comment]
Probably a comment bug. Need to remember how it dammit works actually :).

**Comment by 3Hren** [comment]
What do you think should we do with unit structs/variants? Do nothing? Nil? Or leave serialization as an empty array?

May be make it configurable?

**Comment by SX91** [comment]
@3Hren Haskell's `data-msgpack` [uses Nil representation for units](https://github.com/SX91/msgpack-haskell/blob/master/src/Data/MessagePack/Generic.hs). For unit variant we should use simple `[code]` encoding.
So:
```rust
// Nil
struct Unit
// [code]
enum Enum {
    A,
    B
}
```

I don't think configurations are required here.

**Comment by 3Hren** [comment]
Feels inconsistent. Why not encode unit variant with nil then:

```rust
// [code, nil]
enum Enum {
    A,
    B
}
```

**Comment by SX91** [comment]
But why? Unit struct would be encoded as `Nil` because it's the most compact representation while unit variants would be just `[code]` (the most compact would be `code`, but that one actually _is_ inconsistent). Consider this example:
```rust
enum Enum {
    A,
    B(i32),
    C(i32, i32),
}
// Enum::A - [0]
// Enum::B(0) - [1, 0]
// Enum::C(0, 1) - [2, [0, 1]]
```
This format is compact, unambiguous and separates unit variant, newtype variant and tuple (struct) variant. Structs would be consistent and unambiguous too:
```rust
struct A; // Nil
struct B(i32); // Value
struct C(i32, i32); // [value, value]
```

**Comment by 3Hren** [comment]
Consider:
```rust
enum Enum {
    A, // [0]
    B(i32), // [1, i32]
    C((i32,)), // [2, [i32]]
    D(i32, i32), // [3, [i32, i32]]
    E((i32, i32)), // [4, [i32, i32]]
}
```

I feel this subjectively inconsistent. One can say that it's not. Why not to serialize D(i32, i32) as [3, i32, i32] without inner array and so on.

**Comment by SX91** [comment]
I don't see any inconsistensy here.

> Why not to serialize D(i32, i32) as [3, i32, i32] without inner array and so on.

Newtype variant is a simple value wrapper, while struct/tuple variant is actually a structure. Think of a parser that should
1. choose constructor (variant)
2. parse value
For units it would end up with only choosing a constructor. For others (newtypes and structs/tuples) it would parse value too, which is a value  for newtype variant, map/array for struct/tuple. Serde separates units, newtypes, tuples and structs, so we should just use that.

**Comment by dbrgn** [comment]
I just found out that a newtype struct is serialized as fixarray (0x91). I concur with @SX91 that newtypes are a special case and that they should be transparent. To quote from [Serde docs](https://serde.rs/impl-serialize.html#serializing-a-struct):

> Data formats are encouraged to treat newtype structs as insignificant wrappers around the inner value, serializing just the inner value.

In the meantime, manually implementing `Serialize` and `Deserialize` for the newtype is probably the way to go, right?

**Comment by SX91** [comment]
@dbrgn I'd recommend using master branch for now as those changes are already there.

**Comment by 3Hren** [comment]
Yep, these are on master. I can't currently publish a new release, because I'm in the middle of developing a tool that tries to compile dependent crates and run their tests to figure out what will be broken after publishing (I know about semver, but sudden breaking isn't good). If there are - I feel that I need some kind of migration guide to be done before publishing.

**Comment by dbrgn** [comment]
> I'd recommend using master branch for now as those changes are already there.

Ah, awesome!
